### DISCUSSION TITLE
Unable to reference module/Class for namespace packages.

### DISCUSSION THREAD

**Discussion by RothNRK** [mkdocstrings/mkdocstrings#563] [Q&A] [answered]
I have three namespace packages sitting beside each other. 
```
.
├── Dockerfile
├── LICENSE
├── README.md
├── docker-compose.yml
├── docs
│   ├── ...
├── pyproject.toml
├── src
│   └── mypackage
│       ├── core
│       │   ├── README.md
│       │   ├── mypackage
│       │   │   ├── __init__.py
│       │   │   ├── __main__.py
│       │   │   ├── adapters
│       │   │   │   ├── README.md
│       │   │   │   ├── __init__.py
│       │   │   │   ├── base.py
│       │   │   │   ├── sql.py
│       │   │   │   └── sqlite.py
│       │   │   ├── ...
│       │   │   ├── models
│       │   │   │   ├── __init__.py
│       │   │   │   ├── ...
│       │   │   │   ├── sqlite.py
│       │   │   ├── ...
│       │   ├── setup.cfg
│       │   ├── setup.py
│       ├── mysql
│       │   ├── README.md
│       │   ├── mypackage
│       │   │   ├── __init__.py
│       │   │   ├── adapters
│       │   │   │   ├── __init__.py
│       │   │   │   └── mysql.py
│       │   │   └── models
│       │   │       ├── __init__.py
│       │   │       └── mysql.py
│       │   ├── setup.cfg
│       │   ├── setup.py
│       ├── postgres
│       │   ├── README.md
│       │   ├── mypackage
│       │   │   ├── __init__.py
│       │   │   ├── adapters
│       │   │   │   ├── __init__.py
│       │   │   │   └── postgres.py
│       │   │   └── models
│       │   │       ├── __init__.py
│       │   │       └── postgres.py
│       │   ├── setup.cfg
│       │   ├── setup.py
└── tox.ini
```
If I install the three packages:
```
pip install src/mypackage/core src/mypackage/mysql src/mypackage/postgres

python -c "from mypackage.models import sql" <-- works
python -c "from mypackage.models import mysql" <-- works
python -c "from mypackage.models import postgres" <-- works
```

I'm able to reference the modules/classes
```
::: mypackage.models.sql
::: mypackage.models.mysql
::: mypackage.models.postgres
```

If I install them as editable packages:
```
pip uninstall -y mypackage-core mypackage-mysql mypackage-postgres
pip install -e src/mypackage/core -e src/mypackage/mysql -e src/mypackage/postgres

python -c "from mypackage.models import sqlite" <-- works
python -c "from mypackage.models import mysql" <-- works
python -c "from mypackage.models import postgres" <-- works
```

Will fail while trying to reference the package
```
::: mypackage.models.sqlite  <- good
::: mypackage.models.mysql <- fail
::: mypackage.models.postgres <- fail

ERROR    -  mkdocstrings: mypackage.models.mysql could not be found
ERROR    -  Error reading page 'supported-platforms/mysql.md':
ERROR    -  Could not collect 'mypackage.models.mysql'
```

Trying the same but instead of editable packages, I modify the `PYTHONPATH`

```
export PYTHONPATH="${PYTHONPATH}:$(pwd)/src/mypackage/core"
export PYTHONPATH="${PYTHONPATH}:$(pwd)/src/mypackage/mysql"
export PYTHONPATH="${PYTHONPATH}:$(pwd)/src/mypackage/postgres"

python -c "from mypackage.models import sqlite" <-- works
python -c "from mypackage.models import mysql" <-- works
python -c "from mypackage.models import postgres" <-- works
```

Will fail in the same way that the editable packages do.
```
::: mypackage.models.sqlite  <- good
::: mypackage.models.mysql <- fail
::: mypackage.models.postgres <- fail

ERROR    -  mkdocstrings: mypackage.models.mysql could not be found
ERROR    -  Error reading page 'supported-platforms/mysql.md':
ERROR    -  Could not collect 'mypackage.models.mysql'
```

Simplified mkdocs.yml
```
site_name: mypackage
watch:
  - src/mypackage
  - docs

nav:
  - Supported Platforms:
      - Introduction: supported-platforms/index.md
      - SQLite: supported-platforms/sqlite.md
      - Postgres: supported-platforms/postgres.md
      - MySQL: supported-platforms/mysql.md
 ```

Any idea how I can reference these classes in namespace packages? What have I missed? 

Thank you in advance.

**Comment by pawamoy** [comment]
Hi, is your repo public? I guess you're using pkgutil-style or pkg_resources-style namespace packages, and not native namespace packages? https://packaging.python.org/en/latest/guides/packaging-namespace-packages/

When you install them in editable mode, what do you have in your venv? `.pth` files? `__editable_*.py` modules? both?

Try running mkdocs with the -v option to get the debug logs.

**Reply by RothNRK** [reply]
Yes, I'm using the pkgutil-style package. 

My repo: [blackline-core](https://github.com/blacklinedata/blackline-core)

I run things in a fresh docker container.

Ok, I just ran with `-v` and I see that:

```
DEBUG    -  griffe: Loading path /src/src/blackline/core/blackline/*
...
```
But it only finds the `*-core/*` paths, not the `*-mysql/*` or `*-postgres/*`.

**Reply by pawamoy** [reply, marked as answer]
Thanks. Looking at your `__init__.py` modules, Griffe won't be able to detect your packages as namespace ones, because it supports only the exact code that is documented in https://packaging.python.org/en/latest/guides/packaging-namespace-packages/, i.e.:

```python
__path__ = __import__('pkgutil').extend_path(__path__, __name__)
```

So your code:

```python
from pkgutil import extend_path
from typing import List

__path__: List[str] = extend_path(__path__, __name__)

__version__ = "0.1.0a1"
```

...is not supported unfortunately (and that explains why it finds only `core`: it stops at the first one found). That's because Griffe does not run the code and only parses it. It runs a regular expression to find if the package is a pkgutil/resources-style namespace package. We could handle more code variants (such as yours) by actually visiting the AST, though I'm not sure how much more robust it would be or how easy/hard it would be to implement.

Is using pkgutil a requirement? Maybe you could use native namespace packages instead, by simply removing the `__init__.py` modules?

**Reply by RothNRK** [reply]
Thanks for the info. Just switched it to the implicit namespace package and it looks like it's working as expected. Thank you very much for your help!

**Reply by pawamoy** [reply]
Great! You're welcome :slightly_smiling_face:
